File: backend/src/controllers/submission.controller.js
import Interview from "../models/Interview.js";

export const submitCode = async (req, res) => {
    try {

        const { interviewId, code } = req.body;

        if (!interviewId || !code) {
            return res.status(400).json({
                success: false,
                message: "Interview ID and code are required",
            });
        }

        const interview = await Interview.findById(
            interviewId
        );

        if (!interview) {
            return res.status(404).json({
                success: false,
                message: "Interview not found",
            });
        }

        if (!interview.question) {
            return res.status(400).json({
                success: false,
                message: "Interview question not generated",
            });
        }

        const hiddenTestCases =
            interview.question.hidden_test_cases;

        if (
            !hiddenTestCases ||
            hiddenTestCases.length === 0
        ) {
            return res.status(400).json({
                success: false,
                message: "No test cases available",
            });
        }


        return res.status(200).json({
            success: true,
            message: "Submission received",
            data: {
                interviewId,
                testCases: hiddenTestCases.length,
            },
        });

    } catch (error) {

        return res.status(500).json({
            success: false,
            message: error.message,
        });
    }
};

File: ai-service/app/controllers/submission_controller.py
from fastapi import APIRouter

from app.models.submission import SubmissionRequest
from app.execution.test_case_runner import TestCaseRunner


router = APIRouter(
    prefix="/api/submission",
    tags=["Submission"]
)


runner = TestCaseRunner()


@router.post("/run")
async def run_submission(request: SubmissionRequest):

    if request.language.lower() != "cpp":

        return {
            "success": False,
            "message": "Currently only C++ is supported."
        }

    hidden_test_cases = request.question.get(
        "hidden_test_cases",
        []
    )

    if not hidden_test_cases:

        return {
            "success": False,
            "message": "No hidden test cases found."
        }

    result = runner.run_test_cases(
        code=request.code,
        test_cases=hidden_test_cases
    )

    return {
        "success": True,
        "message": "Code evaluated successfully",
        "data": result
    }